### PR TITLE
Tensor docs

### LINKED ISSUE #62146: torch.tensor、torch.as_tensor do not always use the current device for the default tensor type when device param is none
## 📚 Documentation

image
When input data is a tensor, these ops always infer their device from the tensor.

cc @brianjo @mruberry @gchanan

### PR DESCRIPTION
**Pull request by mruberry**
Fixes #62146.

Modernizes and clarifies the documentation of torch.tensor and torch.as_tensor, highlighting the distinction in their copying behavior and preservation of autograd history.

### CODE REVIEW

File: torch/_torch_docs.py [hunk 1]
```diff
@@ -792,15 +792,29 @@ def merge_dicts(*dicts):
...
+:attr:`dtype` and :attr:`device` then data itself is returned, but if data is a
+tensor with a different dtype or device then it's copied as if using
+`data.to(dtype=dtype, device=device)`.
+
+If data is a NumPy array (an ndarray) with the same dtype and device then a
+tensor is constructed using :func:`torch.from_numpy`.
+
+.. seealso::
+
+    :func:`torch.tensor` never shares its data and creates a new "leaf tensor."
```
Comment (albanD) [inline torch/_torch_docs.py:808]: nit: the dot. should be outside of the quotes?

File: torch/_torch_docs.py [hunk 2]
```diff
@@ -792,15 +792,29 @@ def merge_dicts(*dicts):
            r"""
 as_tensor(data, dtype=None, device=None) -> Tensor
 
-Convert the data into a `torch.Tensor`. If the data is already a `Tensor` with the same `dtype` and `device`,
-no copy will be performed, otherwise a new `Tensor` will be returned with computational graph retained if data
-`Tensor` has ``requires_grad=True``. Similarly, if the data is an ``ndarray`` of the corresponding `dtype` and
-the `device` is the cpu, no copy will be performed.
+Converts :attr:`data` into a tensor, sharing data and preserving autograd
```
Comment (albanD) [inline torch/_torch_docs.py:795]: Why "tensor" and not `torch.Tensor` ?

File: torch/_torch_docs.py [hunk 3]
```diff
@@ -7995,29 +8009,34 @@ def merge_dicts(*dicts):
            r"""
 tensor(data, *, dtype=None, device=None, requires_grad=False, pin_memory=False) -> Tensor
 
-Constructs a tensor with :attr:`data`.
+Constructs a tensor with no autograd history (a "leaf tensor") by copying :attr:`data`.
```
Comment (albanD) [inline torch/_torch_docs.py:8012]: nit: you can link to is_leaf doc if you want to point to a detailed definition of what leaf means
https://pytorch.org/docs/stable/generated/torch.Tensor.is_leaf.html?highlight=leaf#torch.Tensor.is_leaf

File: torch/_torch_docs.py [hunk 4]
```diff
@@ -7995,29 +8009,34 @@ def merge_dicts(*dicts):
...
-    or :func:`torch.Tensor.detach`.
-    If you have a NumPy ``ndarray`` and want to avoid a copy, use
-    :func:`torch.as_tensor`.
+    When working with tensors prefer using :func:`torch.Tensor.clone`,
+    :func:`torch.Tensor.detach`, and :func:`torch.Tensor.requires_grad_` for
+    readability. Letting `t` be a tensor, ``torch.tensor(t)`` is equivalent to
+    ``t.clone().detach()``, and ``torch.tensor(t, requires_grad=True)``
+    is equivalent to ``t.clone().detach().requires_grad_(True)``.
+
+    To change whether a tensor requires grad without copying it use just
```
Comment (albanD) [inline torch/_torch_docs.py:8022]: This paragraph is not as clear as the others.
Maybe something saying that changing if a Tensor requires grad can be done in other ways:
- inplace via requires_grad_
- as a new Tensor that is a view of the original view detach
- as a new Tensor that is not a view as mentioned above via clone().detach()

**Review by albanD — COMMENTED** [review]
Looks quite good!

File: torch/_torch_docs.py (hunk 1 above)
Comment (mruberry) [reply torch/_torch_docs.py:808]: Good question because English is often a confusing nonsense language, so here's what Google says:

```
Commas and periods always go inside the quotation marks in American English; dashes, colons, and semicolons almost always go outside the quotation marks; question marks and exclamation marks sometimes go inside, sometimes stay outside.
```

File: torch/_torch_docs.py (hunk 2 above)
Comment (mruberry) [reply torch/_torch_docs.py:795]: We don't have a firm doc policy about preferring one versus the other in the descriptions, but in this case it might make more sense to be specific about the class.

File: torch/_torch_docs.py (hunk 3 above)
Comment (mruberry) [reply torch/_torch_docs.py:8012]: Good idea

File: torch/_torch_docs.py (hunk 4 above)
Comment (mruberry) [reply torch/_torch_docs.py:8022]: Good point; I'll try to revise it with your thinking

File: torch/_torch_docs.py (hunk 3 above)
Comment (mruberry) [reply torch/_torch_docs.py:8012]: Updated to point to the autograd mechanics note, which I think does a better job of explaining leaf tensors

**Review by albanD — APPROVED** [review]
Thanks for the update!
